daemon/commands.h

```diff
@@ -0,0 +1,62 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#ifndef NETDATA_COMMANDS_H
+#define NETDATA_COMMANDS_H 1
+
+#ifdef _WIN32
+# define PIPENAME "\\\\?\\pipe\\netdata-cli"
+#else
+# define PIPENAME "/tmp/netdata-ipc"
+#endif
+
+#define MAX_COMMAND_LENGTH 4096
+
+typedef enum cmd {
+    CMD_HELP = 0,
+    CMD_RELOAD_HEALTH,
+    CMD_SAVE_DATABASE,
+    CMD_LOG_ROTATE,
+    CMD_EXIT,
+    CMD_FATAL,
+    CMD_TOTAL_COMMANDS
+} cmd_t;
+
+typedef enum cmd_status {
+    CMD_STATUS_SUCCESS = 0,
+    CMD_STATUS_FAILURE,
+    CMD_STATUS_BUSY
+} cmd_status_t;
+
+#define CMD_STATUS_SUCCESS_STR "SUCCESS"
+#define CMD_STATUS_FAILURE_STR "FAILURE"
+#define CMD_STATUS_BUSY_STR "BUSY"
+
+typedef enum cmd_type {
+    CMD_TYPE_EXCLUSIVE = 0, // No other command can run at the same time
+    CMD_TYPE_ORTHOGONAL,    // Other commands are allowed to run concurrently but calls to this command are serialized
+    CMD_TYPE_IDEMPOTENT,    // Any call to any command is allowed at the same time as its execution
```

Comment:
There is no implication about thread-safety. If the command is not re-entrant but idempotent then it's the command's responsibility to protect any critical sections. The command server will allow concurrent execution.

Reply:
'IDEMPOTENT" is not the right term here then, though I'm not sure what would be (`FULLY_CONCURRENT` maybe?).

Something being 'idempotent'  means that you can call it again with the exact same set of arguments without it causing any change in state, provided there has been no other change in state since the last time it was called.  While some of the commands may fit that description (I'd actually hope that as many as possible are, as it makes working with the API much saner), that really says nothing about the concurrency requirements (except that it's easier to make them idempotent the less concurrent they are).

Reply:
It's true what you're saying, the command server does not guarantee that commands of type `CMD_TYPE_IDEMPOTENT` will stop producing side-effects after multiple calls.